File: orchestrator/src/clients/sandbox.rs
use super::uds::UDSClient;
use serde::Deserialize;
use serde_json::json;

pub struct SandboxClient {
    client: UDSClient,
}

impl SandboxClient {
    pub fn new() -> Self {
        Self { client: UDSClient::new("/tmp/ruthen/sandbox.sock") }
    }

    pub async fn execute(&self, cmd: &str) -> Result<String, String> {
        #[derive(Deserialize)]
        struct ExecResult { verdict: String, stdout: String }

        let res: ExecResult = self.client.call("execute", json!({"cmd": cmd, "cwd": "."})).await?;
        Ok(if !res.stdout.is_empty() { res.stdout } else { res.verdict })
    }

    pub async fn set_workspace(&self, path: &str) -> Result<String, String> {
        #[derive(Deserialize)]
        struct WsResult { verdict: String, audit_ref: String }

        let res: WsResult = self.client.call("set_workspace", json!({"path": path})).await?;
        Ok(res.audit_ref)
    }
}

File: UNIT-01/cli/src/daemons/sandbox.ts
// ── Sandbox daemon client ─────────────────────────────────────────────

import { rpcOnce } from './socket.js'

export const SANDBOX_SOCKET = '/tmp/ruthen/sandbox.sock'

export interface ExecuteResult {
  verdict: string
  audit_ref: string
}
export interface WorkspaceResult extends ExecuteResult {}
export interface PolicyResult extends ExecuteResult {}

export class SandboxClient {
  socketPath: string
  constructor(socketPath = SANDBOX_SOCKET) {
    this.socketPath = socketPath
  }

  execute(cmd: string, opts: { allow_network?: boolean; timeout_ms?: number } = {}) {
    return rpcOnce<ExecuteResult & { error?: { code: number; message: string } }>(this.socketPath, 'cage_execute', {
      cmd,
      ...(opts.allow_network !== undefined ? { allow_network: opts.allow_network } : {}),
      ...(opts.timeout_ms ? { timeout_ms: opts.timeout_ms } : {}),
    })
  }
  setWorkspace(path: string) {
    return rpcOnce<WorkspaceResult>(this.socketPath, 'set_workspace', { path })
  }
  setPolicy(policy: {
    enabled?: boolean
    deny_network?: boolean
    ecosystems?: string[]
    allowed_domains?: string[]
    excluded_commands?: string[]
    allow_write_paths?: string[]
    deny_write_paths?: string[]
    deny_read_paths?: string[]
  }) {
    return rpcOnce<PolicyResult>(this.socketPath, 'set_policy', policy)
  }
}
